Add the option to weight variable groups in ClimWIP

lukasbrunner opened the pull request:
**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValTool/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] Give this pull request a descriptive title that can be used as a one line summary in a changelog
-   [x] Make sure your code is composed of functions of no more than 50 lines and uses meaningful names for variables
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Preferably Codacy code quality checks pass, however a few remaining hard to solve Codacy issues are still acceptable. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes

Modified recipe/diagnostic

-   [x] Update documentation for the recipe to the `doc/sphinx/source/recipes` folder
-   [x] Update provenance information if needed
-   [x] Assign the author(s) of the affected recipe(s) as reviewer(s)


Closes #1853

lukasbrunner commented:
This first try runs form me. I'm doing some output checks now. 

@Peter9192: I decided to split the variable group loop in the main function, I hope that this does not create any unforeseen problems. My main intention there was to make sure that every element in the `*_contributions` parameters is included and have if raise an error if the group was not in the pre-processor. But I can definitely also solve this differently if you prefer. 
(In addition I might add that for our work the overlap between performance and independence groups is normally rather small.)
(And it saves reading the observations for the independence only groups)

ruthlorenz reviewed (approved):
I have tested this and the recipes fail when they should and work fine when everything is correct. But I cannot judge if having two loops over the variable_groups has any unforeseen implications.

lukasbrunner commented:
Hey @Peter9192 ! 
I know you are probably still busy but if you have a couple of minutes to look at this it would be great. Maybe this pull request is actually easier to get through as it does not introduce any additional package? Also @ruthlorenz needs the functionality introduced here for some thing else.

Peter9192 commented:
Hey @lukasbrunner I was actually waiting with this one until #1864 was merged, to avoid any conflicts. But if it's completely independent, I can have a look now.

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -345,25 +345,28 @@ def visualize_and_save_performance(performance: 'xr.DataArray', cfg: dict,
 
-def compute_overall_mean(dataset):
+def compute_overall_mean(dataset, contributions):
     """Normalize all variables in a dataset and return their mean."""
```
```suggestion
    """Normalize all variables in a dataset and return their weighted mean.
    
    Relative weights for each variable group are passed via the recipe.
    """
```

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -345,25 +345,28 @@ def visualize_and_save_performance(performance: 'xr.DataArray', cfg: dict,
         median_dim = 'model_ensemble'
     normalized = dataset / dataset.median(dim=median_dim)
 
+    weights = xr.DataArray([contributions[vg] for vg in dataset],
```
```suggestion
    weights = xr.DataArray([contributions[variable_group] for variable_group in dataset],
```

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -345,25 +345,28 @@ def visualize_and_save_performance(performance: 'xr.DataArray', cfg: dict,
         median_dim = 'model_ensemble'
     normalized = dataset / dataset.median(dim=median_dim)
 
+    weights = xr.DataArray([contributions[vg] for vg in dataset],
```
I know you're using 'contributions' to avoid confusion with the 'weights' in the remainder of this script, but in this line you're going from one term to the other anyway. So I would suggest to just use 'weights' inside the scope of this function, and then outside it you can still use 'contributions'.

lukasbrunner replied:
done!

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -422,15 +425,38 @@ def main(cfg):
+        key for key, value in cfg['performance_contributions'].items()
+        if value > 0
+    ]
+    assert len(variable_groups_independence) > 0
```
Is this necessary? I can imagine you can also calculate weights based on performance only. And if you keep it, would it be nice to write a friendly message for if this assertion fails?

lukasbrunner replied:
Good point updated this to allow either performance only or independence only. (I still exclude both having length one which would lead to equal weighting. In principle we could also allow this but it is a bit of a corner case that is also not useful but makes the code more complex. But I don't have strong opinion on this so let me know if you think this should also be allowed)

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -422,15 +425,38 @@ def main(cfg):
+    ]
+    variable_groups_performance = [
+        key for key, value in cfg['performance_contributions'].items()
+        if value > 0
```
Is this check for > 0 really necessary? I guess the main difference is that they will not show up in the intermediate output plots now, is that it? But then, shouldn't we also have a way to visualize the contributions of each variable group in those figures?

lukasbrunner replied:
In my opinion it is cleaner to treat contribution=0 the same way as not including a variable at all. Otherwise this becomes a strange hybrid case of variable groups showing up in the individual plots but not in the mean plot. If we really want all the intermediate plots we can run a dedicated recipe with everything we want, right?

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -422,15 +425,38 @@ def main(cfg):
     independences = {}
 
-    for variable_group in variable_groups:
+    for variable_group in variable_groups_independence:
```
I wonder how much overhead it is to calculate performance/independence for those variable groups that you'll eventually not include in the final weights. If it's not that much, we wouldn't have to break up this loop.

lukasbrunner replied:
I personally find this cleaner, in particular in anticipation of the perfect model test, which will break symmetry between performance and independence even more. In addition it might be nice to be able to have the possibility of using independence variable groups for which no observations are available which should work like this I suspect (otherwise it will at least make it easier if we ever need it)

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -441,41 +467,36 @@ def main(cfg):
 
         model_ancestors.extend(model_data_files)
 
+    model_ancestors = [*{*model_ancestors}]  # make unique
```
Maybe this is a bit more readable?
```suggestion
    model_ancestors = list(set(model_ancestors))  # only keep unique items
```

Peter9192 commented on esmvaltool/recipes/recipe_climwip.yml:
```diff
@@ -87,8 +87,16 @@ diagnostics:
+        independence_contributions:
+          tas_CLIM: .5
+          pr_CLIM: .25
+          psl_CLIM: 0  # equivalent to not setting it
```
should we add 'default=0' in the documentation then?

lukasbrunner replied:
We kind of already say that?! (just the other way around I guess)
"The values give the relative contribution of each group, with 0 being equivalent to not including the group." (sorry did not manage to link the rst file here... :( )

Peter9192 reviewed (changes requested):
Hey @lukasbrunner ,
Personally I would like it if we calculate performance and independence for all variable groups, and then only apply the contributions weighting later, when we accumulate all variable groups. I can even imagine that you would change the relative contributions based on the intermediate results. 

That said, I find your points between brackets quite convincing, so if you prefer to do it like this, that's fine with me. I tested the code and it seems to work fine. You can ignore my comments about the splitted loop then.

lukasbrunner commented:
Hey @Peter9192 Thanks for the review!! I think I addressed all your comments. I was a bit unsure how exactly to handle the complete skipping of either performance or independence weighting. For now I allow either not setting it at all or setting it to empty. I'm not sure it this creates any unforeseen problems but to me it  seemed most intuitive like this. But I'm happy for any feedback.

Peter9192 commented:
@lukasbrunner thanks for the changes. Before I have another look, could you please merge master into this branch? Now that #1864 is merged, we should solve any merge conflicts that might occur.

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -418,19 +439,53 @@ def visualize_and_save_weights(weights: 'xr.DataArray', cfg: dict,
 
+def get_variable_groups(contribution: str, cfg: dict) -> list:
+    """Return a list of variable groups with contributions > 0."""
+    if cfg.get(contribution, None) is not None:
```
None is the default
```suggestion
    if cfg.get(contribution) is not None:
```

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -418,19 +439,53 @@ def visualize_and_save_weights(weights: 'xr.DataArray', cfg: dict,
+    variable_groups_performance = get_variable_groups(
+        'performance_contributions', cfg)
+
+    if (len(variable_groups_independence) == 0 and
+            len(variable_groups_performance) == 0):
```
I think this also works (and is more readable):
```suggestion
    if not (variable_groups_independence or variable_groups_performance):
```

Peter9192 replied:
And here you could then use `contributions_performance` and `contributions_independence`

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -441,41 +496,43 @@ def main(cfg):
-    overall_performance = compute_overall_mean(performance)
-    visualize_and_save_performance(overall_performance, cfg,
-                                   model_ancestors + obs_ancestors)
+    if len(independences) > 0:
```
```suggestion
    if independences:
```

Peter9192 replied:
And here you could also just use `contributions_independence`. I think it's more readable if we're consistent with what we're testing against. And they should always be the same, right?

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -441,41 +496,43 @@ def main(cfg):
+    else:
+        overall_independence = None
+
+    if len(performances) > 0:
```
```suggestion
    if performances:
```

Peter9192 replied:
same

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -418,19 +439,53 @@ def visualize_and_save_weights(weights: 'xr.DataArray', cfg: dict,
     log_provenance(caption, filename_data, cfg, ancestors)
 
 
+def get_variable_groups(contribution: str, cfg: dict) -> list:
+    """Return a list of variable groups with contributions > 0."""
+    if cfg.get(contribution, None) is not None:
+        return [key for key, value in cfg[contribution].items() if value > 0]
+    return []
```
This suggestion only makes sense in combination with some other suggestions. Basically my proposal is to use the contributions dicts as the key flag for deciding all your if statements.

```suggestion
def parse_contributions(metric: str, cfg: dict) -> list:
    """Only retain variable groups with performance or independence contributions > 0."""
    if cfg.get(f'{metric}_contributions') is not None:
        return {key: value for key, value in cfg[contribution].items() if value > 0}
    return {}
```

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -418,19 +439,53 @@ def visualize_and_save_weights(weights: 'xr.DataArray', cfg: dict,
     models, observations = read_metadata(cfg)
 
-    variable_groups = models.keys()
+    variable_groups_independence = get_variable_groups(
```
So instead of using a list to store the variable groups we want to analyse, we can just use the parsed dict directly. That will save us having to import that again later on. 

```suggestion
    contributions_independence = parse_contributions(
```

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -418,19 +439,53 @@ def visualize_and_save_weights(weights: 'xr.DataArray', cfg: dict,
     independences = {}
 
-    for variable_group in variable_groups:
+    for variable_group in variable_groups_independence:
```
and here again

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -441,41 +496,43 @@ def main(cfg):
+        logger.info('Computing overall mean independence')
+        independence = xr.Dataset(independences)
+        overall_independence = compute_overall_mean(
+            independence, cfg['independence_contributions'])
```
And this is where you won't have to parse it a second time

```suggestion
            independence, independence_contributions
```

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -441,41 +496,43 @@ def main(cfg):
+        logger.info('Computing overall mean performance')
+        performance = xr.Dataset(performances)
+        overall_performance = compute_overall_mean(
+            performance, cfg['performance_contributions'])
```
same

Peter9192 reviewed (changes requested):
Hey @lukasbrunner ,
I think your approach is fine, and it works as far as I tested. Since the code changed quite a bit, the review has become bigger than I had expected. But it's considerably improved from the last round, so let's see if we can improve it still a bit more. 

First, I left a couple of small suggestions. But then I came up with an idea and added some more suggestions. The point I'm trying to address is that, in the main function, you now have a couple of if statements that all use different criteria, and you loop over different lists. I think it would be nicer to have a single 'flag' or 'control variable' for those things, and I think you can use the performance_contributions and independence_contributions for that. I hope my idea is clear from the suggestions. If not, let me know.

PS I see the order of my comments is a bit messed up in the conversation view. Suggest to use the review tab to check them.

lukasbrunner commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -7,6 +7,7 @@
 import os
 from collections import defaultdict
 from datetime import datetime
+from typing import Union
```
I was wondering if it is over the top... Although it IS nice to know that None explicitly is allowed. I'm fine with it either way

bouweandela replied:
It's up to @lukasbrunner, I'm fine with whatever works for you. The [`typing`](https://docs.python.org/3/library/typing.html) module is part of the standard library, so there is no need to add it as a dependency. I would recommend running a test with [`mypy`](https://mypy.readthedocs.io/en/stable/getting_started.html#installing-and-running-mypy) to make sure that you got all the types right.

lukasbrunner commented:
Thanks again @Peter9192 ! I find these suggestions super helpful to improve my coding style, highly appreciated! I hope I did not overlook anything I could not accept all you your suggestion directly

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -418,19 +439,54 @@ def visualize_and_save_weights(weights: 'xr.DataArray', cfg: dict,
     independences = {}
 
-    for variable_group in variable_groups:
+    for variable_group in independence_contributions.keys():
```
I think you don't need the `.keys`
```suggestion
    for variable_group in independence_contributions:
```

Peter9192 commented on esmvaltool/diag_scripts/weighting/climwip.py:
```diff
@@ -418,19 +439,54 @@ def visualize_and_save_weights(weights: 'xr.DataArray', cfg: dict,
+
+        model_ancestors.extend(model_data_files)
+
+    for variable_group in performance_contributions.keys():
```
```suggestion
    for variable_group in performance_contributions:
```

Peter9192 reviewed (approved):
Hey @lukasbrunner thanks again, I think you got everything.
1 small suggestion, and 1 more idea (only if you still feel like it):

Would it make sense to parse the sigmas together with the contributions? Then you catch the assertionerror from line 401 earlier in the diagnostic. Also it makes it easier for the reader, because these parsers let them view all the requirements for the recipe settings in one place (especially since these two are tightly connected).

lukasbrunner commented:
Nice suggestion, thx!

Peter9192 commented:
Okay cool, thanks for accepting all my changes. @bouweandela can you please do the bot magic again? I think I'm not allowed to.. or am I?

Peter9192 commented:
@esmvalbot recipe_climwip.yml

lukasbrunner commented:
Okay one more quick fix. If one does not want to use performance weighting it is now possible to:
1) set all contributions to zero
2) not set any contributions but still keep the performance_contributions parameter (commenting out the last three lines below)
3) not set the performance_contributions parameter at all (commenting out all the lines below)
https://github.com/ESMValGroup/ESMValTool/blob/bba6e3e39312f178a3235c6fdc2b55584d342615/esmvaltool/recipes/recipe_climwip.yml#L91-L94

This Is this correct or should one of these cases not be allowed? 

In addition, if either of 1)-3) is true this line can be either commented out or left without a value:
https://github.com/ESMValGroup/ESMValTool/blob/bba6e3e39312f178a3235c6fdc2b55584d342615/esmvaltool/recipes/recipe_climwip.yml#L90

bouweandela commented:
@esmvalbot recipe_climwip.yml

Peter9192 commented:
We're on a roll! :muscle:

Peter9192 commented:
@bouweandela ping :bell:
